blog/2023-02-17-plugins.md

```diff
@@ -0,0 +1,127 @@
+
+## Plugins In Action
+
+No blog post on a new feature would be complete without a walk-through :) So lets walk through the process of writing our own plugin. Not just any plugin though, but  one that will help our favorite underappreciated platform engineer, Susan. As more teams have started adopting Wing in her organization she has realized that she can make use of plugins to help teams meet requirements for deploying applications into the cloud, without asking them to rewrite their code.
+
+She has identified a common use case where her organization has deployed an IAM role into every AWS Account using nested stacks. The existence of this role as a permission boundary for all IAM roles is enforced through AWS organization SCPs. Thus, without it no IAM role can be created in the account, this is a cause of friction for teams that want to get their Wing applications deployed quickly. (whew thats a whole lot of things a developer should not have to care about)
+
+She has decided to write a plugin that implements 2 hooks `preSynth` and `validate`. During the `preSynth` hook, she wants to add the required permission boundary to all IAM roles in the construct tree. Then she intends to validate the existence of the permission boundary on all roles during the `validate` hook. This way teams can know their app will fail to deploy at compile time rather than deploy, making for faster feedback loops.
+
+Susan starts with writing the bare necessities of a plugin. She creates a file named `permission-boundary-compliance.js` and adds the following code:
+
+```js
+// Add permission boundary to all IAM roles
+exports.preSynth = function (app) { }
+
+// Validate that all IAM roles have a permission boundary
+exports.validate = function (config) { }
+```
+
+She plans to make use of a concept from [CDKTF](https://developer.hashicorp.com/terraform/cdktf) known as [Aspects](https://developer.hashicorp.com/terraform/cdktf/concepts/aspects) to traverse the construct tree and add the permission boundary. She can safely use this since she knows her intended target will be Terraform for AWS.
+
+```js
+var iam_role = require("@cdktf/provider-aws/lib/iam-role");
+var cdktf = require("cdktf");
+
+class PermissionBoundaryAspect {
+  constructor(permissionBoundaryArn) {
+    this.permissionBoundaryArn = permissionBoundaryArn;
+  }
+
+  visit(node) {
+    if (node instanceof iam_role.IamRole) {
+      node.permissionsBoundary = this.permissionBoundaryArn;
+    }
+  }
+}
+
+// Add permission boundary to all IAM roles
+exports.preSynth = function (app) {
+  if (!process.env.PERMISSION_BOUNDARY_ARN) {throw new Error("env var PERMISSION_BOUNDARY_ARN not set")}
+  cdktf.Aspects.of(app).add(new PermissionBoundaryAspect(process.env.PERMISSION_BOUNDARY_ARN))
+}
+
+// Validate that all IAM roles have a permission boundary
+exports.validate = function (config) { }
```

Comment:
I guess `validate` should be left out of this code snippet as it appears in the next one.